labs/a10.md

```diff
@@ -0,0 +1,371 @@
+chunk of data before running the image. This is the image of the container. You can view all of the images you've
+downloaded with the `docker images` command:
+
+```
+baisang@rapture ~/d/labs> docker images
+REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
+fedora              latest              9110ae7f579f        2 weeks ago         235MB
+ubuntu              xenial              f975c5035748        3 weeks ago         112MB
+```
+
+Images can take up quite a bit of space on your machine, so you may want to clean up images that you don't plan on
+using. This is especially relevant if you get errors about not having enough disk space on your machine:
+
+`docker rmi <image_id>`
+
+The image files, as well as various filesystems of containers, are stored in `/var/lib/docker`
+
+We'll go over more commands later on in the lab.
+
+### Dockerfiles
+A more powerful way to interface with Docker is by using a Dockerfile. A Dockerfile allows you to define an image by
+specifying all of the commands you would type manually to create an image. Docker can then build images from a specified
+Dockerfile. These Dockerfiles can be put into version control and the images distributed as a binary to keep track of
+both how the image is constructed and also to keep pre-built images around.
+
+Dockerfiles are very powerful and have many different commands and features. We'll go over a basic example, but you
+should check out the [reference page](https://docs.docker.com/engine/reference/builder/) if you are trying to do anything more
+complex.
+
+Here is an example Dockerfile that will build an image that has `python3.6` installed.
+It will also run `python3.6` directly, so you'll be at a python prompt instead of a bash prompt when you run it.
+
+``` dockerfile
+FROM ubuntu:bionic
+
+RUN apt-get update && apt-get install -y python3.6 --no-install-recommends
+
+CMD ["/usr/bin/python3.6", "-i"]
+```
+
+Note: there are some "best practices" for writing Dockerfiles that the above example doesn't use,
+because it's a basic example. For instance, we probably would want to delete `/var/lib/apt/lists/*`, where `apt`
+stores the package list information from `apt update`, after we are done installing packages. We may also choose to use
+Linux variants that are smaller and lighter, e.g. Alpine Linux. The general philosophy is
+containers should be kept as small and "light" as possible. If you're interested in this stuff, [check out this
+article](https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/).
+
+What is this doing? We specify a base image `ubuntu:bionic` (release 18.04 of ubuntu, the latest LTS as of writing).
+We then specify that we should run (`RUN`) the command `apt-get update` and then `apt-get install python3.6` so we can
+install `python3.6` (note that this package probably isn't in apt on your own VMs!).
```

Comment:
Maybe explain since they run Debian stable they have older package versions than 18.04 will by default? This might not be the place to go into that, but could be of interest.